src/nvim/decoration.c

```diff
@@ -850,8 +859,56 @@ int sign_item_cmp(const void *p1, const void *p2)
   return 0;
 }
 
-static const uint32_t sign_filter[4] = {[kMTMetaSignText] = kMTFilterSelect,
-                                        [kMTMetaSignHL] = kMTFilterSelect };
+static const uint32_t conceal_filter[kMTMetaCount] = {[kMTMetaConcealLines] = kMTFilterSelect };
+
+bool decor_conceal_line(win_T *wp, int row, bool check_cursor)
+{
+  if (wp->w_p_cole < 2
+      || (!check_cursor && ((row + 1 == wp->w_cursor.lnum) && !conceal_cursor_line(wp)))) {
+    return false;
+  }
+
+  size_t keys = wp->w_buffer->b_marktree->n_keys;
+  if (!buf_meta_total(wp->w_buffer, kMTMetaConcealLines)) {
+    decor_providers_invoke_conceal_line(wp, row);
+    return wp->w_buffer->b_marktree->n_keys > keys;
+  }
```

Comment:
This assumes that there is only one source of concealed lines to be checked.

Which is fine for now as it is a private API, but maybe worth of a TODO comment.

Reply:
How? I can't fully check/remember the code right now but doesn't the invoke call do the callback for all sources? The only assumption I see here is that those sources exclusively add `conceal_lines` marks.

Reply:
All sources might not add conceals at the same time. so if source A adds a conceal at line 1 and source B wants to add one at the call for line 10, that later call would be skipped. not a problem in the MVP but worthy of a TODO comment.

Reply:
Still don't see the problem; there is another decor provider conceal line invoke at line 899 below. This first one is only an extra early return if there are no conceal line marks in the buffer at all yet.

From memory it goes like this for every line that is checked:
1. Are there no conceal line marks? Then invoke callback and take increase in marks to mean the line is concealed.
2. If there are conceal line marks Check the mark tree for any marks on the checked line.
3. If no conceal line marks found invoke the callback and again take increase in marks to mean the line is concealed.

I'll check again when I get back home if I can understand what kind of TODO note is warranted, unless you can specify contrary to the above.

Reply:
So this entire block is rather an optimization as an early exit? then _that_ could be a good comment to write.

Reply:
Added some comments in https://github.com/neovim/neovim/compare/23597cda3a1d230c8f1a6e0582968d0272778691..9b504a74e2127c3a528fb843e2f685d30d7d09bb.